### PR TITLE
[Model] Replace embedding models with pooling adapter

### LINKED ISSUE #10674: [RFC]: Make any vLLM model a pooling model
### Motivation.

Currently, we have to open new PRs to add pooling functionality for existing architectures supported in vLLM. Since the code involved is basically the same for each model, there is potential to automate away this boilerplate.

### Proposed Change.

Implement a pooling adapter that can be applied to any existing text generation model in vLLM. To preserve features such as LoRA, PP and multimodality, the adapter simply creates a new subclass of the original model.

- [x] Embedding model adapter: #10769
- [ ] Classification model adapter #11469
- [ ] Reward model adapter #11469

The pooling adapter to apply depends on the purpose of the model. To facilitate this, the `embedding` task will be split up, so that the user can specify which adapter to apply to the model via `--task`:

- [x] #10820
- [x] #11457

Meanwhile, current embedding-related classes will be renamed to avoid confusion between `embed` and other pooling tasks:

- [x] #10801

### Feedback Period.

1-2 weeks

### CC List.

@youkaichao @mgoin @robertgshaw2-neuralmagic @maxdebayser

### Any Other Things.

Note that we can still directly map to pooling models in the model registry. This is used when the model architecture has different pooling defaults (e.g. `pooling_type=CLS` for BERT) or additional modules (e.g. `score` in `Qwen2ForRewardModel`). For models that already support pooling, the adapter returns the original model without modifications.

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by DarkLight1337**
This PR implements the adapter function in #10674 for embedding models. This function is called at startup time (in `get_model_architecture`) if `--task embedding` is passed.

### CODE REVIEW

File: vllm/model_executor/models/adapters.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import torch
+import torch.nn as nn
+
+from .interfaces_base import VllmModelForEmbedding, is_embedding_model
+
+_T = TypeVar("_T", bound=type[nn.Module])
+
+
+def for_embedding(cls: _T) -> _T:
+    """Subclass an existing vLLM model to support embeddings."""
```
Comment (Isotr0py) [inline vllm/model_executor/models/adapters.py:13]:
```suggestion
def embedding_model_wrapper(cls: _T) -> _T:
    """Wrapper function to convert an existing vLLM model to embedding model"""
```
I think `for_embedding` might be a little bit confusing. We can use a clearer name for this.

File: vllm/model_executor/models/adapters.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            weights = ((name, data) for name, data in weights
+                       if not name.startswith("lm_head."))
+
+            # If `*ForCausalLM` defines `load_weights` on the inner model
+            # and there are no other inner modules with parameters,
+            # we support loading from both `*Model` and `*ForCausalLM`
+            if (hasattr(self, "model") and hasattr(self.model, "load_weights")
+                    and all(name == "model" or all(False
+                                                   for _ in child.parameters())
+                            for name, child in self.named_children())):
```
Comment (Isotr0py) [inline vllm/model_executor/models/adapters.py:75]: Is there a way to improve the readability of this condition? Maybe we can handle the condition of `*Model` and `*ForCausalLM` separately?

File: vllm/model_executor/models/adapters.py (hunk 1 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/adapters.py:13]: Renamed it to `as_embedding_model`. I'm hesitant to use "wrapper" because it doesn't really wrap the original class - instead it creates a new subclass.

File: vllm/model_executor/models/adapters.py (hunk 2 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/adapters.py]: See if it's more readable now.

Comment (Isotr0py) [reply vllm/model_executor/models/adapters.py]: Nice, it looks better now!

File: vllm/model_executor/models/adapters.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            vllm_config: "VllmConfig",
+            prefix: str = "",
+            **kwargs: Any,
+        ) -> None:
+            super().__init__(vllm_config=vllm_config, prefix=prefix, **kwargs)
+
+            # These are not used in embedding models
+            for attr in ("lm_head", "logits_processor"):
+                if hasattr(self, attr):
+                    delattr(self, attr)
```
Comment (Isotr0py) [inline vllm/model_executor/models/adapters.py:40]:
```suggestion
            for attr in ("lm_head", "logits_processor", "sampler"):
                if hasattr(self, attr):
                    delattr(self, attr)
```
We can remove sampler as well, right?

Comment (DarkLight1337) [reply vllm/model_executor/models/adapters.py:40]: I skipped sampler because it's an immutable property of some multi-modal models.

Comment (Isotr0py) [reply vllm/model_executor/models/adapters.py:40]: OK, make sense!

**Review by Isotr0py — APPROVED** [review]
Overall LGTM!
